design/tls-client-verification.md

```diff
@@ -0,0 +1,231 @@
+# Client certificate validation (mTLS)
+
+Status: Draft
+
+## Goals
+
+- Allow client certificate validation (mTLS) on contour /performed by Envoy)
+- Allow various ways for client certificate validation; Spki, Hash and CA
+- Forward client certificate details to the backend (in the XFCC header)
+- Document mTLS configuration and client certificate detail forwarding.
+
+## Non Goals
+
+- Support for k8s ingress documents.
+- Repeat configuration details described in Envoy documentation (make references)
+
+## Background
+
+Client certificate validation (mTLS) is supported by Envoy. It should
+be possible for `contour` users to utilize this feature.
+
+
+## Implementation phases
+
+The most urgent and also the most straight-forward feature is to
+provide client certificate validation. For some backend applications
+(but likely not for all) the client cerificate details are needed. So
+the proposal is to divide the mTLS implementation in two phases;
+
+1. Add client certificate validation but with hard-coded client
+   certification detail forwarding (XFCC header)
+
+2. Add configuration of what is to be forwarded in the XFCC header
+
+Configuration of the XFCC headed can be added later without make any
+backward-incompatible changes.
+
+The main problem with XFCC is that it is not really part of the tls
+configuration, but rather a backend http connection setting. This
+makes me believe that some discussions must be conducted before the
+configuration is settled. But in the meanwhile implementation the more
+urgent mTLS validation must not be stalled.
+
+The only one of the "other" ingress controllers that mention client
+certificate details forwarding I have found is `traefik`;
+
+https://docs.traefik.io/v2.0/middlewares/passtlsclientcert/
+
+For some reason they do not use the "standard" XFCC header. Never the
+less is may be wise to mimic this approach in `contour`, for instance
+herd-code forwaring of the entire certificate rather than the hash in
+phase 1.
+
+
+## High-Level Design
+
+At a high level I propose the following:
+
+1. A new record "clientValidation" is added in spec.virtualhost.tls
+   in the IngressRoute.
+
+2. The `clientValidation` contains configuration for client
+   certificate validation. Many ways of client certificate validation
+   may be specified (all ways supported by Envoy).
+
+3. For phase 1 the configuration of what to forward in the XFCC header is
+   hard-coded to SANITIZE_SET and certificate hash.
+
+The design of phase 2, the XFCC configuration, is dependent on the
```

Comment:
I'd like to see an issue for something like XFCC should be supported by Contour. If that's the actual goal of this proposal then lets agree on that first before getting side tracked into implementation issues.

Reply:
Please see the comment above